squiRL/common/utils.py

```diff
@@ -0,0 +1,59 @@
+import numpy as np
+import torch
+from torch.utils.data._utils import collate
+from squiRL.common.data_stream import Experience
+from collections import defaultdict
+
+
+def collate_episodes(batch):
+    """Manually processes collected batch of experience
+
+    Args:
+        batch (TYPE): Current mini batch of replay data
+
+    Returns:
+        TYPE: Processed mini batch of replay data
+    """
+
+    # extract from batch
+    batch_dict = {
+        j: [s[i] for s in batch][0]
+        for i, j in enumerate(Experience._fields)
+    }
+
+    # remove streams with no full episodes
```

Comment:
You can bypass all this if you split the episodes before saving them to the buffer, while skipping incomplete ones.

Reply:
Would you mind elaborating further? Im adding experience to the buffer step-wise, not episode-wise. So there's no way for me to know that an episode is incomplete before adding experience to the buffer. I don't want to change this because what if we have agents that learn per step rather than per episode.

Reply:
Right now you are adding stuff step-wise, that's correct, however, it's yielded episode wise. Therefore, we could make use of the fact that we only yield once we have x episodes stored, by just storing full episode rollouts.

Reply:
If you have agents that learn per step then the rollout collector class would change anyway. The collate function shouldn't handle things like this in my opinion.

Reply:
Done. All the work no longer happens here. All data is already in the form of episodes. Also, I am now using your `reset_all()` function which means we dont need to discard the begginings of data streams, since the envs get reset when the policy updates. We still remove trailing ends but no begginings.